src/layers/output.cpp

```diff
@@ -44,8 +44,12 @@ namespace mlp {
     b_ = graph_->param(name + "_b", {1, numOutputClasses}, inits::zeros());
 
   /*const*/ int lemmaDimEmb = options_->get<int>("lemma-dim-emb", 0);
+  std::string lemmaDependency = options_->get<std::string>("lemma-dependency", "");
   ABORT_IF(lemmaDimEmb && !factoredVocab_, "--lemma-dim-emb requires a factored vocabulary");
-  if(lemmaDimEmb > 0) {  // > 0 means to embed the (expected) word with a different embedding matrix
+  if(lemmaDependency == "re-embedding") {  // embed the (expected) word with a different embedding matrix
+    ABORT_IF(
+        lemmaDimEmb <= 0,
+        "In order to predict factors by re-embedding them, a lemma-dim-emb must be specified.");
 #define HARDMAX_HACK
```

Comment:
This was done for what I believe was quick experimentation by who wrote initially this code. Basically if you used a odd number in `--lemma-dim-emb` you could easily change between what is now the options `re-embedding` and `hard-transformer-layer` set through `--lemma-dependency`. Actually I think I can remove this part of the code as it is so experimental. What do you think? I'm suggesting removing: 
https://github.com/marian-nmt/marian-dev/blob/9fa166be885b025711f27b35453e0f2c00c9933e/src/layers/output.cpp#L49-L52
and
https://github.com/marian-nmt/marian-dev/blob/9fa166be885b025711f27b35453e0f2c00c9933e/src/layers/output.cpp#L246-L255

Reply:
If the `hard-transformer-layer` implementation is the same, it can be removed. Though, in order to preserve the old API, when fed an odd integer via`--lemma-dim-emb` (without specifying `--lemma-dependency`), it should select `hard-transformer-layer` with `lemma-dim-emb` adjusted to the desired even number. This should be accompanied by a warning stating this behaviour and that the user _really_ wants an odd number they should also specify `--lemma-dependency`.

Reply:
I see your point and agree with you that the old API should be preserved at all costs to guarantee backwards compatibility. Although for this particular case I think it could be deleted. Here
https://github.com/marian-nmt/marian-dev/blob/9fa166be885b025711f27b35453e0f2c00c9933e/src/layers/output.cpp#L246-L248
You can see that that is a message saying that this was done for quick experimentation. And also, choosing hard max be setting a odd value in `lemma-dim-emb` does not seem a clean solution to me. As @snukky said here https://github.com/marian-nmt/marian-dev/pull/748#discussion_r643878326 we should try to delete this experimental code, and I do think that in this case there's no need to guarantee backwards compatibility for such an experimental scenario

Reply:
If the plan is to not support this experimental mode, then it can be deleted with no associated changes

Reply:
Can I remove this section @snukky? Can we delete this experimental mode?

Reply:
I think so. Let's remove this and I will re-run internal regression tests again before merging this PR to make sure it was only an experimental code.

Reply:
Thanks to both for the inputs and I also agree that deleting and rerunning regressions tests is the best approach. I've deleted the code.

Reply:
Internal regression tests seem to work, I will re-run all regression tests now. Thanks!